frontend/src/components/widgets/options/ToolOptions.vue

```diff
@@ -38,29 +43,39 @@ const wasm = import("@/../wasm/pkg");
 export default defineComponent({
 	props: {
 		activeTool: { type: String },
+		currentToolOptions: { type: Object },
 	},
-	computed: {},
-	methods: {
-		async setShapeOptions(newValue: number) {
-			// TODO: Each value-input widget (i.e. not a button) should map to a field in an options struct,
-			// and updating a widget should send the whole updated struct to the backend.
-			// Later, it could send a single-field update to the backend.
-
-			// This is a placeholder call, using the Shape tool as an example
-			// eslint-disable-next-line camelcase
-			(await wasm).set_tool_options(this.$props.activeTool || "", { Shape: { shape_type: { Polygon: { vertices: newValue } } } });
-		},
-		async setLineOptions(newValue: number) {
-			// eslint-disable-next-line camelcase
-			(await wasm).set_tool_options(this.$props.activeTool || "", { Line: { weight: newValue } });
+	computed: {
+		activeToolOptions(): Record<string, object> {
+			return (this.$props.currentToolOptions || {})[this.$props.activeTool || ""];
 		},
-		async setPenOptions(newValue: number) {
-			// eslint-disable-next-line camelcase
-			(await wasm).set_tool_options(this.$props.activeTool || "", { Pen: { weight: newValue } });
+	},
+	methods: {
+		async updateToolOptions(path: string[], newValue: number) {
+			this.setToolOption(path, newValue);
+			(await wasm).set_tool_options(this.$props.activeTool || "", this.activeToolOptions);
 		},
 		async sendToolMessage(message: string | object) {
 			(await wasm).send_tool_message(this.$props.activeTool || "", message);
 		},
+		// Traverses the given path into the active tool's option struct, and sets the value at the path tail
+		setToolOption(path: string[], newValue: number) {
+			let value = this.activeToolOptions as Record<string, object | number>;
+			[this.$props.activeTool || "", ...path.slice(0, -1)].forEach((attr) => {
```

Comment:
Yup, I changed these two functions around to make it much clearer. I think this kind of logic will turn into something more idiomatic once the layout system is on the Rust side.

I don't think the `forEach` could be a `map`, because rather than mapping each attribute within `optionPath` to a different value, it's using `currentRecord` as an accumulator in order to follow the path of nesting. I just tried to make it a `reduce`, but I don't think that works when the accumulator is a different data type than the list being iterated over. I'll keep looking for fancy alternatives, but the I think the `forEach` is a simple option.

Reply:
"Clear" is always better than "fancy". Keep it as a forEach if map doesn't work.